### ISSUE TITLE
Machine Learning Based Indicators

### ISSUE THREAD

**Issue by slavakurilyak** [produvia/kryptos#23]
# User Story

As a machine learning developer, I want to use machine learning or deep learning to predict the direction of any crypto-asset (i.e. btc_usdt), so that I can develop ML-based indicators.

# Consider

- Consider leveraging amicks's [Speculator](https://github.com/amicks/Speculator) library, which currently supports:

1. Deep Neural Network via [tensorflow](https://github.com/tensorflow/tensorflow) library
2. Random Forest via [scikit-learn](https://github.com/scikit-learn/scikit-learn) library

Speculator can be deployed as a web app serving REST API prediction endpoints. For example: `GET: /api/public/predict/?<bool:use_db>&<str:model_type>&<str:symbol>&<str:unit>&<int:count>&<int:period>&<int:partition>&<int:delta>&<int:seed>&<int:trees>&<int:jobs>&<DelimitedList<str>:longs>`

- Consider leveraging maxim5's [time-series-machine-learning](https://github.com/maxim5/time-series-machine-learning) library, which uses currently supports:

1. Ordinary linear model via [numpy](https://github.com/numpy/numpy) library
2. Gradient boosting [xgboost](https://github.com/dmlc/xgboost) library
3. Deep neural network via [tensorflow](https://github.com/tensorflow/tensorflow) library
4. Recurrent neural network: LSTM, GRU, one or multi-layered via [tensorflow](https://github.com/tensorflow/tensorflow) library
5. Convolutional neural network for 1-dimensional data via [tensorflow](https://github.com/tensorflow/tensorflow) library

- Consider using llens's [CryptoCurrencyTrader](https://github.com/llens/CryptoCurrencyTrader) library, which uses currently supports:

_Support Vector Machine (SVM) Models_

1. Epsilon-Support Vector Regression (SVR) via [scikit-learn](https://github.com/scikit-learn/scikit-learn)
2. C-Support Vector Classification (SVC) via [scikit-learn](https://github.com/scikit-learn/scikit-learn)

_Ensemble Models_
1. Random Forest Classifier via [scikit-learn](https://github.com/scikit-learn/scikit-learn)
2. Random Forest Regressor via [scikit-learn](https://github.com/scikit-learn/scikit-learn)
3. Ada Boost Classifier via [scikit-learn](https://github.com/scikit-learn/scikit-learn)
4. Ada Boost Regressor via [scikit-learn](https://github.com/scikit-learn/scikit-learn)
5. Gradient Boosting Classifier via [scikit-learn](https://github.com/scikit-learn/scikit-learn)
6. Gradient Boosting Regressor via [scikit-learn](https://github.com/scikit-learn/scikit-learn)
7. Extra Trees Classifier via [scikit-learn](https://github.com/scikit-learn/scikit-learn)
8. Extra Trees Regressor via [scikit-learn](https://github.com/scikit-learn/scikit-learn)

_Recurrent Neural Network (RNN) Models_

1. Long Short-Term Memory (LSTM) Models  via [tensorflow](https://github.com/tensorflow/tensorflow)

- Consider using machine learning algorithms to learn indicators from unstructured data ([Coates et al., 2011](https://scholar.google.ca/scholar?cluster=16678777530215029568&hl=en&as_sdt=0,5))

- Consider predicting the direction of cryptocurrency prices using deep learning algorithms, such as LSTM (Long Short Term Memory) ([Ordóñez et at., 2016](https://scholar.google.ca/scholar?cluster=9678884816556007035&hl=en&as_sdt=0,5), [Fischer et al., 2017](https://scholar.google.ca/scholar?cluster=6284581449554292603&hl=en&as_sdt=0,5), [Pang et al., 2018](https://scholar.google.ca/scholar?hl=en&as_sdt=0%2C5&q=%22An+innovative+neural+network+approach+for+stock+market+prediction%22&btnG=)).

# Inspiration

- Askmike's [gekko](https://github.com/askmike/gekko) library, which is developing the [integration of machine learning to replace technical analysis](https://github.com/askmike/gekko/issues/789).

- For some research inspiration, check out [Forecasting to Classification: Predicting the direction of
stock market price using Xtreme Gradient Boosting](https://docs.google.com/viewer?url=https%3A%2F%2Fwww.researchgate.net%2Fprofile%2FSnehanshu_Saha%2Fpublication%2F309492895_Forecasting_to_Classification_Predicting_the_direction_of_stock_market_price_using_Xtreme_Gradient_Boosting%2Flinks%2F581384c008aedc7d8961e371.pdf)

- @bukosabino's [btctrading](https://github.com/bukosabino/btctrading) library which explores XGBoost on time-series data
